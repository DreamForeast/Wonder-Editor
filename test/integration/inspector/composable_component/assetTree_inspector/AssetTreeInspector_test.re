open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("AssetTreeInspector", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initStateAndGl(~sandbox, ());
      EventListenerTool.buildFakeDom()
      |> EventListenerTool.stubGetElementByIdReturnFakeDom;
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("prepare currentSelectSource", () => {
      beforeEach(() => {
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorAssetTool.initAssetTree,
        );
        CurrentSelectSourceEditorService.setCurrentSelectSource(
          EditorType.Asset,
        )
        |> StateLogicService.getAndSetEditorState;
      });
      afterEach(() =>
        StateAssetService.getState()
        |> CurrentNodeDataAssetService.clearCurrentNodeData
        |> CurrentNodeParentIdAssetService.clearCurrentNodeParentId
        |> StateAssetService.setState
        |> ignore
      );

      describe("test component snapshot", () => {
        test("if hasn't current node, show nothing", () =>{
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot() |> ignore;

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildEmptyAppState(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch
        }
        );

        describe("else", () => {
          test("test set folder to be current node", () => {
          let assetTreeDomRecord =
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
          let component = BuildComponentTool.buildAssetComponent();

          assetTreeDomRecord
          |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstFolderDomIndexForAssetTree
          |> MainEditorAssetTool.clickAssetTreeNodeToSetCurrentNode(component);

                      BuildComponentTool.buildInspectorComponent(
                        TestTool.buildEmptyAppState(),
                        InspectorTool.buildFakeAllShowComponentConfig(),
                      )
                      |> ReactTestTool.createSnapshotAndMatch
          });

          test("test set texture to be current node", () => {
                  let assetTreeDomRecord =
                    MainEditorAssetTool.buildTwoLayerAssetTreeRoot();

                  assetTreeDomRecord
                  |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstTextureDomIndex
                  |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildEmptyAppState(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });

          test("test set json to be current node", () => {

                  let assetTreeDomRecord =
                    MainEditorAssetTool.buildTwoLayerAssetTreeRoot();

                  assetTreeDomRecord
                  |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstJsonDomIndex
                  |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildEmptyAppState(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
        });
      });

      describe("test node rename", () => {
        let _triggerInspectorRenameEvent = (inspectorComponent, newName) => {
          BaseEventTool.triggerComponentEvent(
            inspectorComponent,
            AssetTreeInspectorTool.triggerRenameChangeEvent(newName),
          );
          BaseEventTool.triggerComponentEvent(
            inspectorComponent,
            AssetTreeInspectorTool.triggerRenameBlurEvent(newName),
          );
        };
        test("test rename to specific name", () => {

          let assetTreeDomRecord =
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
          let component = BuildComponentTool.buildAssetComponent();
          
          assetTreeDomRecord
          |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstFolderDomIndexForAssetTree
          |> MainEditorAssetTool.clickAssetTreeNodeToSetCurrentNode(component);
        let inspectorComponent =
          BuildComponentTool.buildInspectorComponent(
            TestTool.buildEmptyAppState(),
            InspectorTool.buildFakeAllShowComponentConfig(),
          );
          _triggerInspectorRenameEvent(inspectorComponent, "mickeyFolder");

          inspectorComponent |> ReactTestTool.createSnapshotAndMatch;
        });

        describe("test the root folder can't be rename", () =>
          test("the root treeNode->rename-input->disabled should be true", () => {
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot() |> ignore;

            BaseEventTool.triggerComponentEvent(
              BuildComponentTool.buildAssetComponent(),
              AssetTreeEventTool.clickRootAssetTreeNode,
            );

            BuildComponentTool.buildInspectorComponent(
              TestTool.buildEmptyAppState(),
              InspectorTool.buildFakeAllShowComponentConfig(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
        describe("test rename asset tree children node", () =>
          describe("if node has ext name", () => {
            test("rename input shouldn't show it", () => {
                  let assetTreeDomRecord =
                    MainEditorAssetTool.buildTwoLayerAssetTreeRoot();

                  assetTreeDomRecord
                  |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstJsonDomIndex
                  |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

              BuildComponentTool.buildInspectorComponent(
                TestTool.buildEmptyAppState(),
                InspectorTool.buildFakeAllShowComponentConfig(),
              )
              |> ReactTestTool.createSnapshotAndMatch;
            });
            test("if rename success, the newName should include ext name", () => {
                  let assetTreeDomRecord =
                    MainEditorAssetTool.buildTwoLayerAssetTreeRoot();

                  assetTreeDomRecord
                  |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstJsonDomIndex
                  |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;
              
              _triggerInspectorRenameEvent(
                BuildComponentTool.buildInspectorComponent(
                  TestTool.buildEmptyAppState(),
                  InspectorTool.buildFakeAllShowComponentConfig(),
                ),
                "mickey_json",
              );

              BuildComponentTool.buildAssetComponent()
              |> ReactTestTool.createSnapshotAndMatch;
            });
          })
        );
        describe("deal with specific case", () =>
          test(
            "key in '', trigger onBlur, the input value should be original name",
            () => {

          let assetTreeDomRecord =
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
          let component = BuildComponentTool.buildAssetComponent();

          assetTreeDomRecord
          |> MainEditorAssetNodeTool.OperateTwoLayer.getFirstFolderDomIndexForAssetTree
          |> MainEditorAssetTool.clickAssetTreeNodeToSetCurrentNode(component);

            let inspectorComponent =
              BuildComponentTool.buildInspectorComponent(
                TestTool.buildEmptyAppState(),
                InspectorTool.buildFakeAllShowComponentConfig(),
              );
            _triggerInspectorRenameEvent(inspectorComponent, "");

            inspectorComponent |> ReactTestTool.createSnapshotAndMatch;
          })
        );
      });
    });
  });