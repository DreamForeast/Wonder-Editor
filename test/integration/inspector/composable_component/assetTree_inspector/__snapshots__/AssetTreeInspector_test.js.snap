// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetTreeInspector prepare currentSelectSource test component snapshot else test set folder to be current node 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-assetTree"
  >
    <div
      className=""
    >
      <h1>
        Folder
      </h1>
      <hr />
      <span
        className=""
      >
        name:
      </span>
      <input
        className="input-component float-input"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="newFolder"
      />
    </div>
  </article>
</article>
`;

exports[`AssetTreeInspector prepare currentSelectSource test component snapshot else test set json to be current node 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-assetTree"
  >
    <div>
      <h1>
        Json
      </h1>
      <hr />
      <span
        className=""
      >
        name:
      </span>
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="json"
      />
      <p>
        json result
      </p>
    </div>
  </article>
</article>
`;

exports[`AssetTreeInspector prepare currentSelectSource test component snapshot else test set texture to be current node 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-assetTree"
  >
    <article
      className="wonder-texture-assetTree"
    >
      <div
        className=""
      >
        <h1>
          Texture
        </h1>
        <hr />
        <div
          className=""
        >
          <article
            className="wonder-string-input"
          >
            <span
              className="component-label"
            >
              name : 
            </span>
            <input
              className="input-component float-input"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value="texture3"
            />
          </article>
        </div>
        <div
          className=""
        >
          <article
            className="wonder-select"
          >
            <span
              className=""
            >
              WrapS Mode
            </span>
            <select
              onChange={[Function]}
              value="0"
            >
              <option
                value="2"
              >
                REPEAT
              </option>
              <option
                value="1"
              >
                MIRRORED_REPEAT
              </option>
              <option
                value="0"
              >
                CLAMP_TO_EDGE
              </option>
            </select>
          </article>
        </div>
        <div
          className=""
        >
          <article
            className="wonder-select"
          >
            <span
              className=""
            >
              WrapT Mode
            </span>
            <select
              onChange={[Function]}
              value="0"
            >
              <option
                value="2"
              >
                REPEAT
              </option>
              <option
                value="1"
              >
                MIRRORED_REPEAT
              </option>
              <option
                value="0"
              >
                CLAMP_TO_EDGE
              </option>
            </select>
          </article>
        </div>
        <div
          className=""
        >
          <article
            className="wonder-select"
          >
            <span
              className=""
            >
              Mag Filter Mode
            </span>
            <select
              onChange={[Function]}
              value="1"
            >
              <option
                value="0"
              >
                NEAREST
              </option>
              <option
                value="1"
              >
                LINEAR
              </option>
              <option
                value="2"
              >
                NEAREST_MIPMAP_NEAREST
              </option>
              <option
                value="3"
              >
                LINEAR_MIPMAP_NEAREST
              </option>
              <option
                value="4"
              >
                NEAREST_MIPMAP_LINEAR
              </option>
              <option
                value="5"
              >
                LINEAR_MIPMAP_LINEAR
              </option>
            </select>
          </article>
        </div>
        <div
          className=""
        >
          <article
            className="wonder-select"
          >
            <span
              className=""
            >
              Min Filter Mode
            </span>
            <select
              onChange={[Function]}
              value="0"
            >
              <option
                value="0"
              >
                NEAREST
              </option>
              <option
                value="1"
              >
                LINEAR
              </option>
              <option
                value="2"
              >
                NEAREST_MIPMAP_NEAREST
              </option>
              <option
                value="3"
              >
                LINEAR_MIPMAP_NEAREST
              </option>
              <option
                value="4"
              >
                NEAREST_MIPMAP_LINEAR
              </option>
              <option
                value="5"
              >
                LINEAR_MIPMAP_LINEAR
              </option>
            </select>
          </article>
        </div>
      </div>
    </article>
  </article>
</article>
`;

exports[`AssetTreeInspector prepare currentSelectSource test component snapshot if hasn't current node, show nothing 1`] = `
<article
  className="wonder-inspector-component"
/>
`;

exports[`AssetTreeInspector prepare currentSelectSource test node rename deal with specific case key in '', trigger onBlur, the input value should be original name 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-assetTree"
  >
    <div
      className=""
    >
      <h1>
        Folder
      </h1>
      <hr />
      <span
        className=""
      >
        name:
      </span>
      <input
        className="input-component float-input"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="newFolder"
      />
    </div>
  </article>
</article>
`;

exports[`AssetTreeInspector prepare currentSelectSource test node rename test rename asset tree children node if node has ext name if rename success, the newName should include ext name 1`] = `
<article
  className="wonder-asset-component"
>
  <div
    className="asset-tree"
  >
    <article
      className="wonder-asset-header"
    >
      <div
        className="header-item"
      >
        <button
          onClick={[Function]}
        >
          addFolder
        </button>
      </div>
      <div
        className="header-item"
      >
        <button
          disabled={false}
          onClick={[Function]}
        >
          remove
        </button>
      </div>
      <div
        className="header-item"
      >
        <input
          className="file-upload"
          multiple={true}
          onChange={[Function]}
          type="file"
        />
      </div>
    </article>
    <article
      className="wonder-asset-assetTree"
    >
      <ul
        className="wonder-tree-node"
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "background": "#c0c0c0",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          <img
            src="./public/img/12.jpg"
          />
          Asset
        </li>
        <ul
          className="wonder-tree-node"
          draggable={true}
          onDragEnd={[Function]}
          onDragStart={[Function]}
        >
          <li
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid red",
              }
            }
          >
            <div
              className="item-ground"
              draggable={true}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
            />
            <img
              src="./public/img/12.jpg"
            />
            newFolder
          </li>
        </ul>
        <ul
          className="wonder-tree-node"
          draggable={true}
          onDragEnd={[Function]}
          onDragStart={[Function]}
        >
          <li
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid red",
              }
            }
          >
            <div
              className="item-ground"
              draggable={true}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
            />
            <img
              src="./public/img/12.jpg"
            />
            newFolder
          </li>
        </ul>
      </ul>
    </article>
  </div>
  <article
    className="wonder-asset-assetChildren"
  >
    <article
      className="wonder-asset-folderBox"
      id="folder-2"
      style={
        Object {
          "border": "1px solid red",
        }
      }
    >
      <div
        className="item-ground"
        draggable={true}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
      />
      <img
        src="./public/img/11.jpg"
      />
      <span
        className="item-text"
      >
        newFolder
      </span>
    </article>
    <article
      className="wonder-asset-folderBox"
      id="folder-3"
      style={
        Object {
          "border": "1px solid red",
        }
      }
    >
      <div
        className="item-ground"
        draggable={true}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
      />
      <img
        src="./public/img/11.jpg"
      />
      <span
        className="item-text"
      >
        newFolder
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="texture3img"
      />
      <span
        className="item-text"
      >
        texture3
      </span>
    </article>
    <article
      className="wonder-asset-fileBox item-active"
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="./public/img/12.jpg"
      />
      <span
        className="item-text"
      >
        mickey_json.json
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="texture5img"
      />
      <span
        className="item-text"
      >
        texture5
      </span>
    </article>
  </article>
</article>
`;

exports[`AssetTreeInspector prepare currentSelectSource test node rename test rename asset tree children node if node has ext name rename input shouldn't show it 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-assetTree"
  >
    <div>
      <h1>
        Json
      </h1>
      <hr />
      <span
        className=""
      >
        name:
      </span>
      <input
        className="input-component float-input"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="json"
      />
      <p>
        json result
      </p>
    </div>
  </article>
</article>
`;

exports[`AssetTreeInspector prepare currentSelectSource test node rename test rename to specific name 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-assetTree"
  >
    <div
      className=""
    >
      <h1>
        Folder
      </h1>
      <hr />
      <span
        className=""
      >
        name:
      </span>
      <input
        className="input-component float-input"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="mickeyFolder"
      />
    </div>
  </article>
</article>
`;

exports[`AssetTreeInspector prepare currentSelectSource test node rename test the root folder can't be rename the root treeNode->rename-input->disabled should be true 1`] = `
<article
  className="wonder-inspector-component"
>
  <article
    className="wonder-inspector-assetTree"
  >
    <div
      className=""
    >
      <h1>
        Folder
      </h1>
      <hr />
      <span
        className=""
      >
        name:
      </span>
      <input
        className="input-component float-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value="Asset"
      />
    </div>
  </article>
</article>
`;
